### ISSUE TITLE
socker_server 生成的音频刺啦刺啦的听不清

### ISSUE THREAD

**Issue by WGS-note** [SWivid/F5-TTS#567]
### Checks

- [X] This template is only for bug reports, usage problems go with 'Help Wanted'.
- [X] I have thoroughly reviewed the project documentation but couldn't find information to solve my problem.
- [X] I have searched for existing issues, including closed ones, and couldn't find a solution.
- [X] I confirm that I am using English to submit this report in order to facilitate communication.

### Environment Details

torch                         2.3.0
torchaudio                    2.3.0

### Steps to Reproduce

开发者团队你们好，首先非常感谢回复！
我有两个问题：
1. 为什么 `socket_sever `的推理速度很快？
2. 我在 `socket_sever` 中，保存了 `audio_chunk`，但是当我听音频的时候，它刺啦刺啦的听不清楚。

`socket_sever.py` 内容如下：

```python
import socket
import struct
import torch
import torchaudio
from threading import Thread


import gc
import traceback

from f5_tts.infer.utils_infer import infer_batch_process, preprocess_ref_audio_text, load_vocoder, load_model
from f5_tts.model.backbones.dit import DiT


class TTSStreamingProcessor:
    def __init__(self, ckpt_file, vocab_file, ref_audio, ref_text, vocoder_name, vocoder_local_path, device=None, dtype=torch.float32):
        self.device = device or ("cuda" if torch.cuda.is_available() else "cpu")

        # Load the model using the provided checkpoint and vocab files
        self.model = load_model(
            model_cls=DiT,
            model_cfg=dict(dim=1024, depth=22, heads=16, ff_mult=2, text_dim=512, conv_layers=4),
            ckpt_path=ckpt_file,
            mel_spec_type="vocos",  # or "bigvgan" depending on vocoder
            vocab_file=vocab_file,
            ode_method="euler",
            use_ema=True,
            device=self.device,
        ).to(self.device, dtype=dtype)

        # Load the vocoder
        # self.vocoder = load_vocoder(is_local=False)
        self.vocoder_name = vocoder_name
        self.vocoder_local_path = vocoder_local_path
        self.vocoder = load_vocoder(vocoder_name=self.vocoder_name, is_local=True, local_path=self.vocoder_local_path)

        # Set sampling rate for streaming
        self.sampling_rate = 24000  # Consistency with client

        # Set reference audio and text
        self.ref_audio = ref_audio
        self.ref_text = ref_text

        # Warm up the model
        self._warm_up()

    def _warm_up(self):
        """Warm up the model with a dummy input to ensure it's ready for real-time processing."""
        print("Warming up the model...")
        ref_audio, ref_text = preprocess_ref_audio_text(self.ref_audio, self.ref_text)
        audio, sr = torchaudio.load(ref_audio)
        gen_text = "Warm-up text for the model."

        # Pass the vocoder as an argument here
        infer_batch_process((audio, sr), ref_text, [gen_text], self.model, self.vocoder, device=self.device)
        print("Warm-up completed.")

    def generate_stream(self, text, play_steps_in_s=0.5):
        """Generate audio in chunks and yield them in real-time.
        接收一个文本输入，并使用 infer_batch_process() 方法生成音频。音频生成后，分成小块进行流式传输。每个音频块的大小由 play_steps_in_s 控制
        """
        # Preprocess the reference audio and text
        ref_audio, ref_text = preprocess_ref_audio_text(self.ref_audio, self.ref_text)

        # Load reference audio
        audio, sr = torchaudio.load(ref_audio)

        # Run inference for the input text
        audio_chunk, final_sample_rate, _ = infer_batch_process(
            (audio, sr),
            ref_text,
            [text],
            self.model,
            self.vocoder,
            device=self.device,  # Pass vocoder here
        )


        # torchaudio.save("/home/wangguisen/projects/tts/test_audio/output/socket_audio.wav", torch.from_numpy(audio_chunk).unsqueeze(0), final_sample_rate)
        # print("!!! torchaudio.save !!!")


        # Break the generated audio into chunks and send them
        chunk_size = int(final_sample_rate * play_steps_in_s)

        if len(audio_chunk) < chunk_size:
            packed_audio = struct.pack(f"{len(audio_chunk)}f", *audio_chunk)
            yield packed_audio
            return

        for i in range(0, len(audio_chunk), chunk_size):
            chunk = audio_chunk[i : i + chunk_size]

            # Check if it's the final chunk
            if i + chunk_size >= len(audio_chunk):
                chunk = audio_chunk[i:]

            # Send the chunk if it is not empty
            if len(chunk) > 0:
                packed_audio = struct.pack(f"{len(chunk)}f", *chunk)
                yield packed_audio


def handle_client(client_socket, processor):
    try:
        while True:
            # Receive data from the client
            data = client_socket.recv(1024).decode("utf-8")
            if not data:
                break

            try:
                # The client sends the text input
                text = data.strip()

                # Generate and stream audio chunks
                for audio_chunk in processor.generate_stream(text):
                    client_socket.sendall(audio_chunk)

                # Send end-of-audio signal
                client_socket.sendall(b"END_OF_AUDIO")

            except Exception as inner_e:
                print(f"Error during processing: {inner_e}")
                traceback.print_exc()  # Print the full traceback to diagnose the issue
                break

    except Exception as e:
        print(f"Error handling client: {e}")
        traceback.print_exc()
    finally:
        client_socket.close()


def start_server(host, port, processor):
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.bind((host, port))
    server.listen(5)
    print(f"Server listening on {host}:{port}")

    while True:
        client_socket, addr = server.accept()
        print(f"Accepted connection from {addr}")
        client_handler = Thread(target=handle_client, args=(client_socket, processor))
        client_handler.start()


if __name__ == "__main__":
    try:
        # Load the model and vocoder using the provided files
        ckpt_file = "/home/wangguisen/projects/tts/F5-TTS/ckpts/F5TTS_Base/model_1200000.safetensors"
        vocab_file = "/home/wangguisen/projects/tts/F5-TTS/src/f5_tts/infer/examples/vocab.txt"  # Add vocab file path if needed
        ref_audio = "/home/wangguisen/projects/tts/test_audio/wgs-f5tts_test.wav"
        ref_text = "那到时候再给你打电话，麻烦你注意接听"

        vocoder_name = "vocos"
        vocoder_local_path = "/home/wangguisen/projects/tts/F5-TTS/ckpts/vocos-mel-24khz"

        # Initialize the processor with the model and vocoder
        processor = TTSStreamingProcessor(
            ckpt_file=ckpt_file,
            vocab_file=vocab_file,
            ref_audio=ref_audio,
            ref_text=ref_text,
            dtype=torch.float32,
            vocoder_name=vocoder_name,
            vocoder_local_path=vocoder_local_path,
        )

        # Start the server
        start_server("0.0.0.0", 9998, processor)
    except KeyboardInterrupt:
        gc.collect()

```

保存音频的代码注释掉了，如下：

```python
# torchaudio.save("/home/wangguisen/projects/tts/test_audio/output/socket_audio.wav", torch.from_numpy(audio_chunk).unsqueeze(0), final_sample_rate)
# print("!!! torchaudio.save !!!")
```

**我检查了生成audio的极值，发现客户端请求得到的audio和启动服务端Warm-up时的极值不一样，如下所示：**
```shell
    def _warm_up(self):
        """Warm up the model with a dummy input to ensure it's ready for real-time processing."""
        print("Warming up the model...")
        ref_audio, ref_text = preprocess_ref_audio_text(self.ref_audio, self.ref_text)
        audio, sr = torchaudio.load(ref_audio)
        gen_text = "Warm-up text for the model."

        # Pass the vocoder as an argument here
        # infer_batch_process((audio, sr), ref_text, [gen_text], self.model, self.vocoder, device=self.device)
        warm_audio_chunk, warm_final_sample_rate, _  = infer_batch_process((audio, sr), ref_text, [gen_text], self.model, self.vocoder, device=self.device)
        # torchaudio.save("/home/wangguisen/projects/tts/test_audio/output/tttt_scoket.wav", torch.from_numpy(warm_audio_chunk).unsqueeze(0), warm_final_sample_rate)

        print(f"[DEBUG] Warm-up Max audio value: {np.max(warm_audio_chunk)}")
        print(f"[DEBUG] Warm-up Min audio value: {np.min(warm_audio_chunk)}")


        print("Warm-up completed.")

    def generate_stream(self, text, play_steps_in_s=0.5):
        """Generate audio in chunks and yield them in real-time.
        接收一个文本输入，并使用 infer_batch_process() 方法生成音频。音频生成后，分成小块进行流式传输。每个音频块的大小由 play_steps_in_s 控制
        """
        # Preprocess the reference audio and text
        ref_audio, ref_text = preprocess_ref_audio_text(self.ref_audio, self.ref_text)

        # Load reference audio
        audio, sr = torchaudio.load(ref_audio)

        # Run inference for the input text
        audio_chunk, final_sample_rate, _ = infer_batch_process(
            (audio, sr),
            ref_text,
            [text],
            self.model,
            self.vocoder,
            device=self.device,  # Pass vocoder here
        )


        print("???? ", type(audio_chunk), audio_chunk.shape, final_sample_rate)
        print(f"[DEBUG] Max audio value: {np.max(audio_chunk)}")
        print(f"[DEBUG] Min audio value: {np.min(audio_chunk)}")
        # torchaudio.save("/home/wangguisen/projects/tts/test_audio/output/socket_audio.wav", torch.from_numpy(audio_chunk).unsqueeze(0), final_sample_rate)
```
`Warm-up`时，生成音频的极值为：
```python
[DEBUG] Warm-up Max audio value: 0.6296077966690063
[DEBUG] Warm-up Min audio value: -0.948524534702301
```

`请求服务端时`，生成音频的极值为：
```python
[DEBUG] Max audio value: 0.48660799860954285
[DEBUG] Min audio value: -0.5944586396217346
```

然而，我的客户端代码是使用的 https://github.com/SWivid/F5-TTS/blob/main/src/f5_tts/infer/README.md 这里面的啊

### ✔️ Expected Behavior

在 `socket_sever` 中，将音频保存，合成的效果能够和 `infer_cli` 一样

### ❌ Actual Behavior

在 `socket_sever` 中，保存了 `audio_chunk`，但是当我听音频的时候，它刺啦刺啦的听不清楚。

**Comment by WGS-note** [comment]
我已经解决！客户端代码如下：

```python
async def listen_to_voice(text, server_ip='localhost', server_port=9999, rate=24000, output_file="./out.wav"):
    client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    client_socket.connect((server_ip, server_port))

    audio_data = []

    async def play_audio_stream():
        buffer = b''
        p = pyaudio.PyAudio()
        stream = p.open(format=pyaudio.paFloat32,
                        channels=1,
                        rate=24000,  # Ensure this matches the server's sampling rate
                        output=True,
                        frames_per_buffer=2048)

        try:
            while True:
                chunk = await asyncio.get_event_loop().run_in_executor(None, client_socket.recv, 1024)
                if not chunk:  # End of stream
                    break
                if b"END_OF_AUDIO" in chunk:
                    buffer += chunk.replace(b"END_OF_AUDIO", b"")
                    if buffer:
                        audio_array = np.frombuffer(buffer, dtype=np.float32).copy()  # Make a writable copy
                        stream.write(audio_array.tobytes())

                        audio_data.append(audio_array)
                    break
                buffer += chunk
                if len(buffer) >= 4096:
                    audio_array = np.frombuffer(buffer[:4096], dtype=np.float32).copy()  # Make a writable copy
                    stream.write(audio_array.tobytes())
                    buffer = buffer[4096:]

                    audio_data.append(audio_array)
        finally:
            stream.stop_stream()
            stream.close()
            p.terminate()

    try:
        # Send only the text to the server
        await asyncio.get_event_loop().run_in_executor(None, client_socket.sendall, text.encode('utf-8'))
        await play_audio_stream()
        print("Audio playback finished.")

        audio_data = np.concatenate(audio_data)
        with open(output_file, "wb") as f:
            sf.write(f.name, audio_data, rate)
            print(f.name)

        print(f"[DEBUG] audio_data: {audio_data.shape}  save at {output_file}")


    except Exception as e:
        print(f"Error in listen_to_voice: {e}")

    finally:
        client_socket.close()
```

还有一个问题，能不能加快推理？做到毫秒级别？

**Comment by SWivid** [comment]
great! if you have found the cause, you can make a pr (we are not familiar with the socket.py so that we may not just find the modified part, it was contributed by the community)
for speed up, we have response in the corresponding issue.

很高兴你解决了问题。如果是存在bug的话欢迎pr（我们对这块儿没那么熟悉，是社区帮忙贡献的code，所以可能一下子看不出来哪儿改了）

关于推理加速，用tensorrt啥的优化下能快很多，但这个我们就不太擅长了，同样欢迎pr~

**Comment by AwsomeName** [comment]
> 我已经解决了！客户端代码如下：
> 
> ```python
> async def listen_to_voice(text, server_ip='localhost', server_port=9999, rate=24000, output_file="./out.wav"):
>     client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
>     client_socket.connect((server_ip, server_port))
> 
>     audio_data = []
> 
>     async def play_audio_stream():
>         buffer = b''
>         p = pyaudio.PyAudio()
>         stream = p.open(format=pyaudio.paFloat32,
>                         channels=1,
>                         rate=24000,  # Ensure this matches the server's sampling rate
>                         output=True,
>                         frames_per_buffer=2048)
> 
>         try:
>             while True:
>                 chunk = await asyncio.get_event_loop().run_in_executor(None, client_socket.recv, 1024)
>                 if not chunk:  # End of stream
>                     break
>                 if b"END_OF_AUDIO" in chunk:
>                     buffer += chunk.replace(b"END_OF_AUDIO", b"")
>                     if buffer:
>                         audio_array = np.frombuffer(buffer, dtype=np.float32).copy()  # Make a writable copy
>                         stream.write(audio_array.tobytes())
> 
>                         audio_data.append(audio_array)
>                     break
>                 buffer += chunk
>                 if len(buffer) >= 4096:
>                     audio_array = np.frombuffer(buffer[:4096], dtype=np.float32).copy()  # Make a writable copy
>                     stream.write(audio_array.tobytes())
>                     buffer = buffer[4096:]
> 
>                     audio_data.append(audio_array)
>         finally:
>             stream.stop_stream()
>             stream.close()
>             p.terminate()
> 
>     try:
>         # Send only the text to the server
>         await asyncio.get_event_loop().run_in_executor(None, client_socket.sendall, text.encode('utf-8'))
>         await play_audio_stream()
>         print("Audio playback finished.")
> 
>         audio_data = np.concatenate(audio_data)
>         with open(output_file, "wb") as f:
>             sf.write(f.name, audio_data, rate)
>             print(f.name)
> 
>         print(f"[DEBUG] audio_data: {audio_data.shape}  save at {output_file}")
> 
> 
>     except Exception as e:
>         print(f"Error in listen_to_voice: {e}")
> 
>     finally:
>         client_socket.close()
> ```
> 
> 还有一个问题，不能加速推理？达到毫秒级别？

大哥，咋解决的啊
